### PR TITLE
Change `out_dir` to better reflect the fs needs

### LINKED ISSUE #475: [RFC] don't pass `out_dir` to writers
After an offline discussion with @belltailjp,
he proposes that the path to save objects through the writers should be stored in the filesystem object instead of passing an`out_dir` string directly to the writers.

I agree with this proposal

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #475 

~`out_dir` has now a default value of '.' since using an empty one required major changes.~

I am not happy at all of how this PR turn out to be, I think we may need to rethink how fs and out_dir interact

### CODE REVIEW

**Comment by emcastillo** [comment]
I think it looks a bit better now, but still not 100% happy

@belltailjp , do you have any suggestions 😇 ?

**Comment by belltailjp** [comment]
Thanks so much for the quick prototype!!
This branch worked well for me as expected.

> I think we may need to rethink how fs and out_dir interact

Right... I feel that the responsibility to handle where to save things into is currently a bit diverged (`fs` vs `out_dir` to `Writer.__init__` and `Writer.save`). IMHO the ideal situation would be to integrate them into the `fs` mechanism, so Writer (and extensions) can and should only concern about sub-directories visible through the `fs`, but I agree that it's gonna be a little too much breaking change. In addition doing so would force user to always make FS.
```python
writer = SimpleWriter(PosixFileSystem("/path/to/some/dir"))
```

File: pytorch_pfn_extras/writing/_writer_base.py [hunk 1]
```diff
@@ -89,6 +93,7 @@ def list(
             if path_or_prefix is None:
                 raise ValueError(
                     "'path_or_prefix' must not be None in recursive mode.")
+            path_or_prefix = self.get_actual_path(path_or_prefix)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/writing/_writer_base.py:96]: Need to normalize path even if the `recursive=False` case.

File: pytorch_pfn_extras/writing/_writer_base.py [hunk 2]
```diff
@@ -140,9 +149,11 @@ def makedirs(
...
             exist_ok: bool = False
     ) -> None:
+        file_path = self.get_actual_path(file_path)
         return os.makedirs(file_path, mode, exist_ok)
 
     def exists(self, file_path: str) -> bool:
+        file_path = self.get_actual_path(file_path)
         return os.path.exists(file_path)
 
     def rename(self, src: str, dst: str) -> None:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/writing/_writer_base.py:159]: `get_actual_path` needed?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/writing/_writer_base.py (hunk 1 above)
Comment (emcastillo) [reply pytorch_pfn_extras/writing/_writer_base.py]: fixed!

File: pytorch_pfn_extras/writing/_writer_base.py (hunk 2 above)
Comment (emcastillo) [reply pytorch_pfn_extras/writing/_writer_base.py]: good catch, old code

**Comment by emcastillo** [comment]
PTAL 😄

File: pytorch_pfn_extras/writing/_writer_base.py [hunk 3]
```diff
@@ -89,11 +89,11 @@ def list(
             path_or_prefix: Optional[str] = None,
             recursive: bool = False,
     ) -> Iterator[str]:
+        path_or_prefix = self.get_actual_path(path_or_prefix)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/writing/_writer_base.py]: `get_actual_path` could be None here 😅

Comment (emcastillo) [reply pytorch_pfn_extras/writing/_writer_base.py]: yeah, I think this should fix it, `get_actual_path` fails and `os.scandir` admits `None`

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**